extensions/stats/plugin.cc

```diff
@@ -190,60 +186,89 @@ const std::vector<MetricTag>& PluginRootContext::defaultTags() {
+                    },
+                    static_cast<uint32_t>(Protocol::GRPC), count_peer_labels,
+                    true},
+      MetricFactory{"response_messages_total", MetricType::Counter,
+                    [](::Wasm::Common::RequestInfo& request_info) -> uint64_t {
+                      uint64_t out = request_info.response_message_count -
+                                     request_info.last_response_message_count;
+                      request_info.last_response_message_count =
+                          request_info.response_message_count;
+                      return out;
+                    },
+                    static_cast<uint32_t>(Protocol::GRPC), count_peer_labels,
+                    true},
+
       // TCP metrics.
       MetricFactory{"tcp_sent_bytes_total", MetricType::Counter,
-                    [](const ::Wasm::Common::RequestInfo& request_info)
-                        -> uint64_t { return request_info.tcp_sent_bytes; },
-                    true, count_standard_labels},
+                    [](::Wasm::Common::RequestInfo& request_info) -> uint64_t {
+                      uint64_t out = 0;
+                      std::swap(out, request_info.tcp_sent_bytes);
+                      return out;
+                    },
+                    static_cast<uint32_t>(Protocol::TCP), count_tcp_labels,
+                    true},
       MetricFactory{"tcp_received_bytes_total", MetricType::Counter,
-                    [](const ::Wasm::Common::RequestInfo& request_info)
-                        -> uint64_t { return request_info.tcp_received_bytes; },
-                    true, count_standard_labels},
-      MetricFactory{
-          "tcp_connections_opened_total", MetricType::Counter,
-          [](const ::Wasm::Common::RequestInfo& request_info) -> uint64_t {
-            return request_info.tcp_connections_opened;
-          },
-          true, count_standard_labels},
-      MetricFactory{
-          "tcp_connections_closed_total", MetricType::Counter,
-          [](const ::Wasm::Common::RequestInfo& request_info) -> uint64_t {
-            return request_info.tcp_connections_closed;
-          },
-          true, count_standard_labels},
+                    [](::Wasm::Common::RequestInfo& request_info) -> uint64_t {
+                      uint64_t out = 0;
+                      std::swap(out, request_info.tcp_received_bytes);
+                      return out;
+                    },
+                    static_cast<uint32_t>(Protocol::TCP), count_tcp_labels,
+                    true},
+      MetricFactory{"tcp_connections_opened_total", MetricType::Counter,
+                    [](::Wasm::Common::RequestInfo& request_info) -> uint64_t {
+                      uint8_t out = 0;
+                      std::swap(out, request_info.tcp_connections_opened);
+                      return out;
+                    },
+                    static_cast<uint32_t>(Protocol::TCP), count_tcp_labels,
+                    true},
```

Comment:
curious: why is `connections_opened_total` defined as `recurrent==true` and `connections_closed_total` defined as `recurrent==false` ?

Reply:
Connection opened is reported on timer, but connection closed is only set at the end in onLog.